platform/utm_mgrs_utils.cpp

```diff
@@ -0,0 +1,567 @@
+
+#include <string>
+#include <iostream>
+#include <cmath>
+
+namespace utm_mgrs_utils
+{
+using namespace std;
+
+typedef struct UTMPoint_Value
+{
+  double easting;
+  double northing;
+  int zone_number;
+  char zone_letter;
+
+} UTMPoint;
+
+UTMPoint latlon_to_utm(double lat, double lon);
+int latlon_to_zone_number(double lat, double lon);
+char latitude_to_zone_letter(double lat);
+int zone_number_to_central_longitude(int zone_number);
+double mod_angle(double value);
+
+string utm_to_str(UTMPoint point);
+string utm_to_mgrs_str(UTMPoint point, int precision);
+int zone_to_100k(int i);
+string get_100k_id(double easting, double northing, int zone_number);
+double square_char_to_easting(char e, int set);
+double square_char_to_northing(char n, int set);
+double zone_to_min_northing(char zoneLetter);
+
+const double PI = 3.14159165358979;
+const double K0 = 0.9996;
+const double E = 0.00669438;
+const double E2 = E * E;
+const double E3 = E2 * E;
+const double E_P2 = E / (1 - E);
+const double R = 6378137.0;
+
+const double SQRT_E = sqrt(1 - E);
+const double _E = (1 - SQRT_E) / (1 + SQRT_E);
+const double _E2 = _E * _E;
+const double _E3 = _E2 * _E;
+const double _E4 = _E3 * _E;
+const double _E5 = _E4 * _E;
+
+const double M1 = (1.0 - E / 4.0 - 3.0 * E2 / 64.0 - 5.0 * E3 / 256.0);
+const double M2 = (3.0 * E / 8.0 + 3.0 * E2 / 32.0 + 45.0 * E3 / 1024.0);
+const double M3 = (15.0 * E2 / 256.0 + 45.0 * E3 / 1024.0);
+const double M4 = (35.0 * E3 / 3072.0);
+
+const double P2 = (3.0 / 2.0 * _E - 27.0 / 32.0 * _E3 + 269.0 / 512.0 * _E5);
+const double P3 = (21.0 / 16.0 * _E2 - 55.0 / 32.0 * _E4);
+const double P4 = (151.0 / 96.0 * _E3 - 417.0 / 128.0 * _E5);
+const double P5 = (1097.0 / 512.0 * _E4);
+
+const string ZONE_LETTERS = "CDEFGHJKLMNPQRSTUVWXX";
+double const DEGREES_TO_RADIANS = 0.017453292519943295;
+double const RADIANS_TO_DEGREES = 57.29577951308232;
+
+const int NUM_100K_SETS = 6;
+const int SET_ORIGIN_COLUMN_LETTERS[] = { 'A', 'J', 'S', 'A', 'J', 'S' };
+const int SET_ORIGIN_ROW_LETTERS[] = { 'A', 'F', 'A', 'F', 'A', 'F' };
+
+
+double degToRad(double deg)
+{
+  return deg * DEGREES_TO_RADIANS;
+}
+
+double radToDeg(double rad)
+{
+  return rad * RADIANS_TO_DEGREES;
+}
+
+// Convert lat,lon for WSG 84 ellipsoid to MGRS string.
+string FormatMGRS(double lat, double lon, int precision)
+{
+  if (precision > 5)
+    precision = 5;
+  else if (precision < 1)
+    precision = 1;
+
+  if (lat <= -80 || lat > 84)
+    return "Latitude limit exceeded";
+  if (lon <= -180 || lon > 180)
+    return "Longitude limit exceeded";
+
+  UTMPoint mgrsp = latlon_to_utm(lat, lon);
+
+  // Need to add this to set the right letter for the latitude.
+  mgrsp.zone_letter = latitude_to_zone_letter(lat);
+  return utm_to_mgrs_str(mgrsp, precision);
+}
+
+// Convert lat,lon for WSG 84 ellipsoid to UTM string.
+string FormatUTM(double lat, double lon)
+{
+  if (lat <= -80 || lat > 84)
+    return "Latitude limit exceeded";
```

Comment:
Why `std::optional<string>` can't be used here and in other similar places?

Reply:
Added `std::optional<string>` to all functions and JNI calls.